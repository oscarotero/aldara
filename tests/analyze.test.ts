import analyze, { mergeDefaults } from "../mod.ts";
import { defaults } from "./types.ts";
import { assertSnapshot } from "https://deno.land/std@0.200.0/testing/snapshot.ts";

Deno.test("analyze", async (t) => {
  const schema = await analyze(import.meta.resolve("./types.ts"), {
    private: true,
  });
  mergeDefaults(schema.Blog, defaults);
  await assertSnapshot(t, schema);
});

Deno.test("analyze remote", async (t) => {
  const schema = await analyze(
    "https://deno.land/x/lume@v1.18.4/plugins/esbuild.ts",
  );
  await assertSnapshot(t, schema);
});
