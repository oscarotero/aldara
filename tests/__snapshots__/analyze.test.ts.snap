export const snapshot = {};

snapshot[`analyze 1`] = `
{
  Blog: {
    children: {
      authors: {
        children: {
          children: {
            email: {
              optional: false,
              type: "string",
            },
            name: {
              optional: false,
              type: "string",
            },
          },
          type: "object",
        },
        default: [],
        optional: false,
        type: "array",
      },
      content: {
        default: "Hello World",
        optional: false,
        type: "string",
      },
      date: {
        default: 2020-01-01T00:00:00.000Z,
        optional: false,
        readonly: true,
        type: "object",
        typeName: "Date",
      },
      people: {
        children: [
          {
            type: "string",
          },
          {
            children: {
              children: {
                email: {
                  optional: false,
                  type: "string",
                },
                name: {
                  optional: false,
                  type: "string",
                },
              },
              type: "object",
              typeName: "People",
            },
            type: "array",
          },
        ],
        default: [],
        optional: false,
        type: "union",
      },
      subtitle: {
        default: "subtitle",
        optional: true,
        type: "string",
      },
      tags: {
        children: {
          type: "string",
        },
        default: [],
        optional: false,
        type: "array",
      },
      title: {
        default: "Hello World",
        description: "Hello",
        optional: false,
        readonly: true,
        see: "https://example.com",
        type: "string",
      },
    },
    type: "object",
    typeName: "Blog",
  },
}
`;

snapshot[`analyze remote 1`] = `
{
  EsmOptions: {
    children: {
      cjsExports: {
        children: {
          key: {
            type: "string",
          },
          value: {
            children: [
              {
                children: {
                  type: "string",
                },
                type: "array",
              },
              {
                type: "string",
              },
            ],
            type: "union",
          },
        },
        description: "Configure the cjs-exports option for each package",
        optional: true,
        type: "record",
      },
      deps: {
        children: {
          key: {
            type: "string",
          },
          value: {
            children: [
              {
                children: {
                  type: "string",
                },
                type: "array",
              },
              {
                type: "string",
              },
            ],
            type: "union",
          },
        },
        description: "Configure the deps for each package",
        optional: true,
        type: "record",
      },
      dev: {
        description: "To include the ?dev option to all packages",
        optional: true,
        type: "boolean",
      },
    },
    type: "object",
    typeName: "EsmOptions",
  },
  Options: {
    children: {
      esm: {
        children: {
          cjsExports: {
            children: {
              key: {
                type: "string",
              },
              value: {
                children: [
                  {
                    children: {
                      type: "string",
                    },
                    type: "array",
                  },
                  {
                    type: "string",
                  },
                ],
                type: "union",
              },
            },
            description: "Configure the cjs-exports option for each package",
            optional: true,
            type: "record",
          },
          deps: {
            children: {
              key: {
                type: "string",
              },
              value: {
                children: [
                  {
                    children: {
                      type: "string",
                    },
                    type: "array",
                  },
                  {
                    type: "string",
                  },
                ],
                type: "union",
              },
            },
            description: "Configure the deps for each package",
            optional: true,
            type: "record",
          },
          dev: {
            description: "To include the ?dev option to all packages",
            optional: true,
            type: "boolean",
          },
        },
        description: "Global options for esm.sh CDN used to fetch NPM packages",
        optional: false,
        see: "https://esm.sh/#docs",
        type: "object",
        typeName: "EsmOptions",
      },
      extensions: {
        children: {
          type: "string",
        },
        description: "The list of extensions this plugin applies to",
        optional: false,
        type: "array",
      },
      options: {
        children: {
          absWorkingDir: {
            description: "Documentation: https://esbuild.github.io/api/#working-directory",
            optional: true,
            type: "string",
          },
          alias: {
            children: {
              key: {
                type: "string",
              },
              value: {
                type: "string",
              },
            },
            description: "Documentation: https://esbuild.github.io/api/#alias",
            optional: true,
            type: "record",
          },
          allowOverwrite: {
            description: "Documentation: https://esbuild.github.io/api/#allow-overwrite",
            optional: true,
            type: "boolean",
          },
          assetNames: {
            description: "Documentation: https://esbuild.github.io/api/#asset-names",
            optional: true,
            type: "string",
          },
          banner: {
            children: {},
            description: "Documentation: https://esbuild.github.io/api/#banner",
            optional: true,
            type: "object",
          },
          bundle: {
            description: "Documentation: https://esbuild.github.io/api/#bundle",
            optional: true,
            type: "boolean",
          },
          chunkNames: {
            description: "Documentation: https://esbuild.github.io/api/#chunk-names",
            optional: true,
            type: "string",
          },
          conditions: {
            children: {
              type: "string",
            },
            description: "Documentation: https://esbuild.github.io/api/#conditions",
            optional: true,
            type: "array",
          },
          entryNames: {
            description: "Documentation: https://esbuild.github.io/api/#entry-names",
            optional: true,
            type: "string",
          },
          entryPoints: {
            children: [
              {
                children: {
                  type: "string",
                },
                type: "array",
              },
              {
                children: {
                  key: {
                    type: "string",
                  },
                  value: {
                    type: "string",
                  },
                },
                type: "record",
              },
              {
                children: {
                  children: {
                    in: {
                      optional: false,
                      type: "string",
                    },
                    out: {
                      optional: false,
                      type: "string",
                    },
                  },
                  type: "object",
                },
                type: "array",
              },
            ],
            description: "Documentation: https://esbuild.github.io/api/#entry-points",
            optional: true,
            type: "union",
          },
          external: {
            children: {
              type: "string",
            },
            description: "Documentation: https://esbuild.github.io/api/#external",
            optional: true,
            type: "array",
          },
          footer: {
            children: {},
            description: "Documentation: https://esbuild.github.io/api/#footer",
            optional: true,
            type: "object",
          },
          inject: {
            children: {
              type: "string",
            },
            description: "Documentation: https://esbuild.github.io/api/#inject",
            optional: true,
            type: "array",
          },
          loader: {
            children: {},
            description: "Documentation: https://esbuild.github.io/api/#loader",
            optional: true,
            type: "object",
          },
          mainFields: {
            children: {
              type: "string",
            },
            description: "Documentation: https://esbuild.github.io/api/#main-fields",
            optional: true,
            type: "array",
          },
          metafile: {
            description: "Documentation: https://esbuild.github.io/api/#metafile",
            optional: true,
            type: "boolean",
          },
          nodePaths: {
            children: {
              type: "string",
            },
            description: "Documentation: https://esbuild.github.io/api/#node-paths",
            optional: true,
            type: "array",
          },
          outExtension: {
            children: {},
            description: "Documentation: https://esbuild.github.io/api/#out-extension",
            optional: true,
            type: "object",
          },
          outbase: {
            description: "Documentation: https://esbuild.github.io/api/#outbase",
            optional: true,
            type: "string",
          },
          outdir: {
            description: "Documentation: https://esbuild.github.io/api/#outdir",
            optional: true,
            type: "string",
          },
          outfile: {
            description: "Documentation: https://esbuild.github.io/api/#outfile",
            optional: true,
            type: "string",
          },
          packages: {
            description: "Documentation: https://esbuild.github.io/api/#packages",
            optional: true,
            type: "string",
            value: "external",
          },
          plugins: {
            children: {
              children: {
                name: {
                  optional: false,
                  type: "string",
                },
                setup: {
                  optional: false,
                  type: "function",
                },
              },
              type: "object",
              typeName: "Plugin",
            },
            description: "Documentation: https://esbuild.github.io/plugins/",
            optional: true,
            type: "array",
          },
          preserveSymlinks: {
            description: "Documentation: https://esbuild.github.io/api/#preserve-symlinks",
            optional: true,
            type: "boolean",
          },
          publicPath: {
            description: "Documentation: https://esbuild.github.io/api/#public-path",
            optional: true,
            type: "string",
          },
          resolveExtensions: {
            children: {
              type: "string",
            },
            description: "Documentation: https://esbuild.github.io/api/#resolve-extensions",
            optional: true,
            type: "array",
          },
          splitting: {
            description: "Documentation: https://esbuild.github.io/api/#splitting",
            optional: true,
            type: "boolean",
          },
          stdin: {
            children: {
              contents: {
                children: [
                  {
                    type: "string",
                  },
                  {
                    type: "object",
                    typeName: "Uint8Array",
                  },
                ],
                optional: false,
                type: "union",
              },
              loader: {
                children: [
                  {
                    type: "string",
                    value: "base64",
                  },
                  {
                    type: "string",
                    value: "binary",
                  },
                  {
                    type: "string",
                    value: "copy",
                  },
                  {
                    type: "string",
                    value: "css",
                  },
                  {
                    type: "string",
                    value: "dataurl",
                  },
                  {
                    type: "string",
                    value: "default",
                  },
                  {
                    type: "string",
                    value: "empty",
                  },
                  {
                    type: "string",
                    value: "file",
                  },
                  {
                    type: "string",
                    value: "js",
                  },
                  {
                    type: "string",
                    value: "json",
                  },
                  {
                    type: "string",
                    value: "jsx",
                  },
                  {
                    type: "string",
                    value: "local-css",
                  },
                  {
                    type: "string",
                    value: "text",
                  },
                  {
                    type: "string",
                    value: "ts",
                  },
                  {
                    type: "string",
                    value: "tsx",
                  },
                ],
                optional: true,
                type: "union",
              },
              resolveDir: {
                optional: true,
                type: "string",
              },
              sourcefile: {
                optional: true,
                type: "string",
              },
            },
            description: "Documentation: https://esbuild.github.io/api/#stdin",
            optional: true,
            type: "object",
            typeName: "StdinOptions",
          },
          tsconfig: {
            description: "Documentation: https://esbuild.github.io/api/#tsconfig",
            optional: true,
            type: "string",
          },
          write: {
            description: "Documentation: https://esbuild.github.io/api/#write",
            optional: true,
            type: "boolean",
          },
        },
        description: "The options for esbuild",
        optional: false,
        see: "https://esbuild.github.io/api/#general-options",
        type: "object",
        typeName: "BuildOptions",
      },
    },
    type: "object",
    typeName: "Options",
  },
}
`;
